explorer/interpreter/type_checker.cpp

```diff
@@ -659,6 +660,20 @@ auto TypeChecker::ArgumentDeduction(
   switch (param->kind()) {
     case Value::Kind::VariableType: {
       const auto& var_type = cast<VariableType>(*param);
+      const Value* substituted_param_type = Substitute(deduced, param);
+      const auto& binding = cast<VariableType>(*param).binding();
+      if (binding.has_static_type()) {
+        if (!IsTypeOfType(&binding.static_type())) {
+          if (!IsImplicitlyConvertible(substituted_param_type, arg, impl_scope,
+                                       false)) {
+            return CompilationError(source_loc)
+                   << "cannot convert deduced value " << *arg << " for "
+                   << binding.name() << " to parameter type "
+                   << *substituted_param_type;
+          }
+        }
+      }
```

Comment:
```suggestion
      const auto& binding = cast<VariableType>(*param).binding();
      if (binding.has_static_type()) {
        const Value* binding_type = Substitute(deduced, &binding.static_type());
        if (!IsTypeOfType(binding_type)) {
          if (!IsImplicitlyConvertible(arg, binding_type, impl_scope,
                                       false)) {
            return CompilationError(source_loc)
                   << "cannot convert deduced value " << *arg << " for parameter "
                   << binding.name() << " to parameter type "
                   << *binding_type;
          }
        }
      }
```

Two things here:
1) I got the argument order to `IsImplicitlyConvertible` backwards in a previous suggested edit, sorry about that.
2) I think we want to substitute into the type of the binding, not into `param`. Eg, if we have `[T:! Type, U:! ImplicitAs(T)](x: T, y: U)` then when checking the deduced argument for `U`, we want to see if it's implicitly convertible to `ImplicitAs(T)` after substituting in the deduced value for `T`. Substituting into `param` itself will result in either `param` or whatever value we already deduced for `U`.